### PR TITLE
Format error message for unmatched signature between _out and base functions

### LINKED ISSUE #33547: Bad error message when _out and non-out variants don't match
Keyword arguments are less likely to cause backwards compatibility problems in the future, but we don't make it easily possible to define them. According to @kurtamohler the "obvious" thing doesn't work in https://github.com/pytorch/pytorch/pull/33430#discussion_r381565565

with error

```
Writing torch/csrc/autograd/generated/python_variable_methods.cpp
Traceback (most recent call last):
  File "tools/setup_helpers/generate_code.py", line 93, in <module>
    main()
  File "tools/setup_helpers/generate_code.py", line 88, in main
    options.selected_op_list_path,
  File "tools/setup_helpers/generate_code.py", line 43, in generate_code
    gen_autograd_python(declarations_path or DECLARATIONS_PATH, autograd_gen_dir, 'tools/autograd')
  File "/home/kurtamohler/development/pytorch-bmm/tools/autograd/gen_autograd.py", line 234, in gen_autograd_python
    out, aten_decls + deprecated, template_path)
  File "/home/kurtamohler/development/pytorch-bmm/tools/autograd/gen_python_functions.py", line 173, in gen_py_torch_functions
    env = create_python_bindings(py_torch_functions, is_python_method=False, is_module=False)
  File "/home/kurtamohler/development/pytorch-bmm/tools/autograd/gen_python_functions.py", line 192, in create_python_bindings
    py_methods.append(method_impl(name, overload_decls, is_python_method, is_module))
  File "/home/kurtamohler/development/pytorch-bmm/tools/autograd/gen_python_functions.py", line 884, in method_impl
    grouped = group_overloads(declarations, is_python_method)
  File "/home/kurtamohler/development/pytorch-bmm/tools/autograd/gen_python_functions.py", line 1046, in group_overloads
    x, list(dictionary.keys())))
RuntimeError: 'base' not in dictionary for bmm(Tensor input, Tensor mat2, bool deterministic). keys are ['out', 'signature']

```

### PR DESCRIPTION
**Pull request by ejguan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47087 Format error message for unmatched signature between _out and base functions**

Updated Error Message:
image

Fixes #33547

Differential Revision: [D24633077](https://our.internmc.facebook.com/intern/diff/D24633077)

### CODE REVIEW

File: tools/autograd/gen_python_functions.py [hunk 1]
```diff
@@ -702,8 +702,8 @@ def group_overloads(declarations, is_python_method):
     for x, dictionary in sorted(grouped.items()):
         if 'base' not in dictionary:
             raise RuntimeError(
-                "'base' not in dictionary for {}. keys are {}".format(
-                    x, list(dictionary.keys())))
+                "_out variant function {} can not find matched non-out 'base' function "
+                "{} in dictionary.".format(dictionary['signature'], x))
```
Comment (zou3519) [inline tools/autograd/gen_python_functions.py:706]: Let's try to make this more human readable:
```
While identifying overloads, we found an out= schema {out_schema} without a corresponding non-out variant.
We expected the non-out variant to have schema {expected_schema}. Please check that you spelled the schema correctly in native_functions.yaml.
```

I think the `declarations` argument to gen_python_functions might have possible candidates. If so, we can make the error message even more prescriptive:

```
While identifying overloads, we found an out= schema {out_schema} without a corresponding non-out variant.
We expected the non-out variant to have schema {expected_schema}. Please check that you spelled the schema correctly in native_functions.yaml. We discovered the following candidates:
- {candidate1}
- {candidate2}
...
```

Comment (ejguan) [reply tools/autograd/gen_python_functions.py:706]: > I think the `declarations` argument to gen_python_functions might have possible candidates. If so, we can make the error message even more prescriptive:

Yes, it's doable. And the overhead for a whole loop through `declarations` to find the non-out function is minimum.

**Comment by ejguan** [comment]
Update error message

File: tools/autograd/gen_python_functions.py [hunk 2]
```diff
@@ -701,9 +701,17 @@ def group_overloads(declarations, is_python_method):
...
+                if declaration['name'] == non_out_name and not declaration['deprecated']:
+                    signature = get_python_signature(declaration, is_python_method, skip_outputs=True)
+                    candidates.append(signature)
             raise RuntimeError(
-                "'base' not in dictionary for {}. keys are {}".format(
-                    x, list(dictionary.keys())))
+                "While identifying overloads, we found an out schema {} without a corresponding non-out variant. "
+                "We expected the non-out variant to have schema: \n- {}\nPlease check that you spelled the schema "
+                "correctly or in native_functions.yaml. We discovered the following candidate(s): \n"
+                .format(dictionary['signature'], x) + "\n".join("- {}".format(candidate) for candidate in candidates))
```
Comment (zou3519) [inline tools/autograd/gen_python_functions.py:714]: Can you update the sample error message in the PR body?

**Review by zou3519 — APPROVED** [review]

File: tools/autograd/gen_python_functions.py (hunk 2 above)
Comment (ejguan) [reply tools/autograd/gen_python_functions.py:714]: Sure. There is actually a typo in the error message, and I will fix it and update PR.

**Pull request closed without merging**